_posts/2018-05-07-is-graphql-the-future.markdown

```diff
@@ -0,0 +1,188 @@
+
+Take this query, for example:
+
+```graphql
+{
+  artwork(id: "andy-warhol-campbells-soup-i-black-bean") {
+    title
+    artist {
+      name
+    }
+  }
+}
+```
+
+Here, we tell the server to look up an Artwork by its slug, and tell us the title. So far, this is just like REST. But we _also_ tell it to find us the Artist for us. Importantly, object fields _must_ be followed up with further queries, and scalar fields _cannot_ be. With that in mind, it’s easy to see that `artwork` and `artist` are object fields, while `title` and `name` are scalar fields. Also note that the fact that there’s also an `artist` root query field actually has nothing to do with the fact that
+
+Usefully, the server’s response to a GraphQL request will directly mirror the shape of the request itself. The result of the query above looks like:
+
+```
+{
+  "data": {
+    "artwork": {
+      "title": "Campbell's Soup I: Black Bean",
+      "artist": {
+        "name": "Andy Warhol"
+      }
+    }
+  }
+}
+```
+
+# GraphQL as a (meta-)scripting language
+
+Let’s dig a little deeper into the scripting language interpretation of GraphQL, because this is the crux of how I think people should think of GraphQL. If I were to guess, I think Facebook:
+
+* …knows this is true. After all, much of the spec is devoted to the execution model of GraphQL
+* …might have backed into this design. It’s well known that they think of their data as a graph.
+* …thinks that this too difficult to explain, and thus, settled on the query language paradigm.
+
+There are a couple reasons GraphQL might not look like a scripting language to you. It didn’t to me, at first! After all, it doesn’t have a linear flow of operations. It doesn’t have a concept of (explicit) variables, other than parameters to the whole document. There are no looping constructs or recursion. Let’s look into each of these.
+
+## Operation flow
+
+It’s true that a GraphQL request doesn’t follow the same vertical sequence of steps model familiar to most programming languages. But sequencing _does_ exist. It’s just represented by calling nested fields of object types, terminating in a scalar field. See this query:
+
+```graphql
+{
+  step1(arg: “something”) {
+    step2 {
+      step3(arg: "something else”) {
+        outputScalar
+      }
+    }
+  }
+```
+
+In a more traditional language, this would look more like:
+
+```
+step1(“something”)
+step2()
+return step3(“something else”)
+```
+
+So, sequencing got a bit more verbose, but it _is_ there. Interestingly, GraphQL reserves vertical stacking for something that’s an afterthought in most languages: _concurrency_. Granted, there’s no way to synchronize concurrent paths of execution, to unify them, but still. I’m not going to quote [the spec](https://facebook.github.io/graphql/October2016/), but search it yourself, and you can find the word “parallel” in there several times. This design is intentional.
```

Comment:
vertical stacking feels weird as a word choice because visually it's represented horizontally

Reply:
Hmm, in this paragraph, I'm trying to describe the vertical stacking, but I think it may be confusing because the example above is horizontal. I'll try to clarify this. I might need to add a separate example for the vertical part.